packages/@react-stately/tree/src/useTreeState.ts

```diff
@@ -40,13 +40,15 @@ export interface TreeState<T> {
  * of items from props, item expanded state, and manages multiple selection state.
  */
 export function useTreeState<T extends object>(props: TreeProps<T>): TreeState<T> {
+  let {disallowEmptySelection} = props;
   let [expandedKeys, setExpandedKeys] = useControlledState(
     props.expandedKeys ? new Set(props.expandedKeys) : undefined,
     props.defaultExpandedKeys ? new Set(props.defaultExpandedKeys) : new Set(),
     props.onExpandedChange
   );
 
-  let selectionState = useMultipleSelectionState(props);
+  // If tree disallows empty selection, fire onSelectionChange even if user selects the currently selected option.
+  let selectionState = useMultipleSelectionState({...props, alwaysFireOnSelection: disallowEmptySelection});
```

Comment:
Similar concerns as https://github.com/adobe/react-spectrum/pull/2618/files#r758850888. SideNav has tests where clicking on the already selected SideNav item would fire `onSelectionChange` with the same key if `disallowEmptySelection=true` but that felt kinda weird, perhaps it should support onAction instead? 

Additionally `TreeProps` extends `MultipleSelection` not `MultipleSelectionStateProps` unlike `useListState`/`useGridState`/`useTableState`, not entirely sure if it should be extending `MultipleSelectionStateProps` instead.

Reply:
I'd just remove those tests. SideNav is not yet released.

Reply:
Alright I'll go ahead and add a skip for those tests and remove the above logic. Any thoughts about the `MultipleSelection` vs `MultipleSelectionStateProps` here? I figure we may want to allow users of `useTreeState` to pass `alwaysFireOnSelection` but perhaps we can hold off for now since we don't have a use case in our spectrum components that use `useTreeState`

Reply:
I'd keep it in state props only for now so we don't expose on spectrum components